src/StoreMetaUnpacker.cc

```diff
@@ -10,130 +10,80 @@
-    /*
-     * sanity check on 'buflen' value.  It should be at least big
-     * enough to hold one type and one length.
-     */
-    getBufferLength();
-    if (*hdr_len < MinimumBufferLength)
-        throw TexcHere("store entry metadata is too small");
-    if (*hdr_len > buflen)
-        throw TexcHere("store entry metadata is too big");
+    Assure(fieldStart_ != bufEnd_);
+    fieldStart_ += sizeof(char); // swap meta type
+    fieldStart_ += sizeof(int); // swap meta value length
+    fieldStart_ += meta_.rawLength; // swap meta value
+
+    sync();
+    return *this;
 }
 
+/// (re)set meta_
 void
-StoreMetaUnpacker::getBufferLength()
+Store::SwapMetaIterator::sync()
 {
-    memcpy(hdr_len, &buf[1], sizeof(int));
-}
+    if (fieldStart_ == bufEnd_)
+        return; // nothing to do when we reach the end of iteration
 
-StoreMetaUnpacker::StoreMetaUnpacker(char const *aBuffer, ssize_t aLen, int *anInt) :
-    buf(aBuffer),
-    buflen(aLen),
-    hdr_len(anInt),
-    position(1 + sizeof(int)),
-    type('\0'),
-    length(0),
-    tail(NULL)
-{
-    assert(aBuffer != NULL);
-}
+    // We cannot start beyond the end of the header: We start with valid
+    // begin/end buffer pointers, and each field checks for overreach.
+    Assure(fieldStart_ < bufEnd_);
 
-void
-StoreMetaUnpacker::getType()
-{
-    type = buf[position];
-    ++position;
+    meta_ = SwapMetaView(fieldStart_, bufEnd_);
 }
 
-void
-StoreMetaUnpacker::getLength()
-{
-    memcpy(&length, &buf[position], sizeof(int));
-    position += sizeof(int);
-}
+/* StoreMetaUnpacker */
 
-bool
-StoreMetaUnpacker::doOneEntry()
+StoreMetaUnpacker::StoreMetaUnpacker(const char * const buf, const ssize_t size, int * const swap_hdr_len)
 {
-    getType();
-    getLength();
-
-    if (position + length > *hdr_len) {
-        debugs(20, DBG_CRITICAL, "storeSwapMetaUnpack: overflow!");
-        debugs(20, DBG_CRITICAL, "\ttype=" << type << ", length=" << length << ", *hdr_len=" << *hdr_len << ", offset=" << position);
-        return false;
-    }
-
-    StoreMeta *newNode = StoreMeta::Factory(type, length, &buf[position]);
-
-    if (newNode)
-        tail = StoreMeta::Add (tail, newNode);
-
-    position += length;
+    Assure(buf);
+    Assure(size >= 0);
 
-    return true;
-}
+    // buffer = <metadata> [HTTP response byte]...
+    // metadata = <prefix> [metadata field]...
+    // prefix = <magic> <metadata size a.k.a. swap_hdr_len>
+    // We parse the prefix and then skip it, ready to iterate metadata fields.
 
-bool
-StoreMetaUnpacker::moreToProcess() const
-{
-    return *hdr_len - position - MinimumBufferLength >= 0;
-}
+    const auto requiredPrefixSize = sizeof(char) + sizeof(int);
+    Assure2(uint64_t(size) >= requiredPrefixSize, "parsing buffer accommodates metadata prefix");
 
-StoreMeta *
-StoreMetaUnpacker::createStoreMeta ()
-{
-    tlv *TLV = NULL;
-    tail = &TLV;
-    assert(hdr_len != NULL);
+    if (buf[0] != static_cast<char>(STORE_META_OK))
```

Comment:
With conversion of SwapMetaType to enum class type char this casting should no longer be necessary.
```suggestion
    if (buf[0] != STORE_META_OK)
```

Reply:
Unfortunately, this would not help - enum class requires the explicit casting regardless of the underlying enum type.

Reply:
Ha! You were both tricked by the wrong (and deceiving) name of the `STORE_META_OK` macro. The name has nothing to do with the SwapMetaType enum. This change is slightly out of scope, but I fixed it in branch commit 5247769.